lib/matplotlib/cbook/__init__.py

```diff
@@ -1416,7 +1416,14 @@ def _reshape_2D(X, name):
     *name* is used to generate the error message for invalid inputs.
     """
     # Iterate over columns for ndarrays, over rows otherwise.
-    X = np.atleast_1d(X.T if isinstance(X, np.ndarray) else np.asarray(X))
+    # Delete masked points for MaskedArray
+    if isinstance(X, np.ndarray):
+        X = X.T
+    elif len(X) and isinstance(next(iter(X), None), np.ma.MaskedArray):
+        X = np.asarray([delete_masked_points(x).pop() for x in X])
```

Comment:
Can this not simply be `X = x[~x.mask]`?

Reply:
I've been trying to get that to work, but when I feed the reproduction from #13533 without masking anything like so:
```
...
#a = np.ma.masked_less( a, -9000 )
#b = np.ma.masked_less( b, -9000 )
#c = np.ma.masked_less( c, -9000 )
...
```
The filtered X causes `_reshape_2D`  to throw an exception, namely this one
https://github.com/matplotlib/matplotlib/blob/4c48547629c06bacffb60316f4be1a80b698da39/lib/matplotlib/cbook/__init__.py#L1436

Something about `X.ndim` is not consistent with the change in this particular scenario. Using `delete_masked_points` gave it a value of 2, but `X = x[~x.mask]` gave it a value of 3.
```
(Pdb) X
[masked_array(data=[  229.5,   374. ,   536.5, -9999. , -9999. , -9999. ,
                   -9999. , -9999. ,   182. ,    42.5,    49. ],
             mask=False,
       fill_value=1e+20), masked_array(data=[  363. ,   118.5,   159. , -9999. ,   311. ,   516. ,
                     380. ,   338.5,   223. ,   211.5,   128.5],
             mask=False,
       fill_value=1e+20), masked_array(data=[205.5, 277.5, 141.5, 278. , 302. , 251. , 299. , 250. ,
                   315.5,  92.1, 211.9],
             mask=False,
       fill_value=1e+20)]
(Pdb) np.atleast_1d(np.asarray([delete_masked_points(x).pop() for x in X]))
array([[  229.5,   374. ,   536.5, -9999. , -9999. , -9999. , -9999. ,
        -9999. ,   182. ,    42.5,    49. ],
       [  363. ,   118.5,   159. , -9999. ,   311. ,   516. ,   380. ,
          338.5,   223. ,   211.5,   128.5],
       [  205.5,   277.5,   141.5,   278. ,   302. ,   251. ,   299. ,
          250. ,   315.5,    92.1,   211.9]])
(Pdb) np.atleast_1d(np.asarray([delete_masked_points(x).pop() for x in X])).ndim
2
(Pdb) np.atleast_1d([x[~x.mask] for x in X])
array([[[  229.5,   374. ,   536.5, -9999. , -9999. , -9999. , -9999. ,
         -9999. ,   182. ,    42.5,    49. ]],

       [[  363. ,   118.5,   159. , -9999. ,   311. ,   516. ,   380. ,
           338.5,   223. ,   211.5,   128.5]],

       [[  205.5,   277.5,   141.5,   278. ,   302. ,   251. ,   299. ,
           250. ,   315.5,    92.1,   211.9]]])
(Pdb) np.atleast_1d([x[~x.mask] for x in X]).ndim
3
```

Reply:
Sorry, I mistyped that: `X = X[~X.mask]`

Reply:
Oh, in that case `X` wouldn't have a `mask` attribute, the collections within may have the type `np.ma.MaskedArray` but not itself.

Figured `X = np.asarray([delete_masked_points(x).pop() for x in X])` can be swapped with `X = [(x[~x.mask] if np.ma.getmask(x).any() else x) for x in X]` if preferred, but masked point can't be removed without iterating over `X` as far as I know.

Reply:
OK, I guess I don't quite understand what case you are looking at. I guess I don't understand what X is for this case.  Can you add a test that is more informative?

Reply:
Sure, added a test case that would fail when doing `X = x[~x.mask]`.

Reply:
OK, I think I am slowly understanding.  

Can we also see a test of this in action with boxplot?  Otherwise I'm not sure you are doing the right thing removing the points instead of just setting them to `np.NaN`.  What happens if `X` has some np.NaN values in it?  If it all works OK, then you can just do something like:
```
X = np.ma.asarray(X)
X[X.mask] = np.NaN
X = np.asarray(X)
```

But anyway, this should get a test of the end-user method, not the private cbook method.

Reply:
@jklymak I've added two test cases to `test_axes.py`, the first is where only one array is masked (so in `_reshape_2D`, X is a `MaskedArray` object) and the other is when multiple arrays are masked (so in `_reshape_2D`, X is an array of `MaskedArray` objects)

Let us know if we should remove the tests in `cbook`.

Reply:
This is looking good.  
  - Can you also test a 2-D masked array?  
  - Can you test an array w/ np.NaN values in it (not masked?)

I'm not sure about the cbook test - are there other private cbook tests?  If so, thats fine.

Note you will want to rebase versus master to get rid of the py3.7 doc-build error.  Back up your branch first if you are at all unsure of how to do a rebase!  https://matplotlib.org/devel/gitwash/index.html

Reply:
... in fact I'm 95% sure a 2-d masked array will not do the right thing here...

Reply:
@jklymak are 2-D masked arrays supported? The helper function `delete_masked_points` doesn't seem to support them. 

```
    Arguments can be in any of 5 categories:

    1) 1-D masked arrays
    2) 1-D ndarrays
    3) ndarrays with more than one dimension
    4) other non-string iterables
    5) anything else
```

When I try and use `delete_masked_points` on a 2-D array, it reaches this point and returns a `ValueError`

```
            if isinstance(x, np.ma.MaskedArray):
                if x.ndim > 1:
                    raise ValueError("Masked arrays must be 1-D")
```

This is the test case I was trying it out on
```
def test_boxplot_masked_array_2d():
    fig, ax = plt.subplots()
    data = [[-1, -1], [1, 3], [3, 6]]
    masked_data = np.ma.masked_less(data, 0)
    ax.boxplot(masked_data)
    fig.canvas.draw()
    assert ax.get_ylim() == (1, 6)
```

Reply:
Yes, masked arrays can be of any dimension.  

I'm pretty sure boxplot and friends take 2-D numpy arrays, so they should take 2-D masked arrays too.  `delete_masked_points` of course can't delete from more than 1-D because you may need to remove more or fewer values from each row/column and that won't be two-d anymore.   Thats why I was suggesting you also think about what happens if a np.array has a NaN, because its pretty easy to change a masked array into an array w/ NaNs at the mask points.

Reply:
Ok, I'm able to get test cases passing for a 2D array using this, 
```
    if isinstance(X, np.ma.MaskedArray):
        X = X[~X.mask] if np.ma.getmask(X).any() else X
    elif isinstance(next(iter(X), None), np.ma.MaskedArray):
        X = [(x[~x.mask] if np.ma.getmask(x).any() else x) for x in X]
```

However, when I try with array that has NaN points, it seems like Y limits get changed.
By using this code I don't see anything when I call `plt.show()`
```
import numpy as np
import matplotlib.pyplot as plt

a = np.ma.array([1, 2, 3, np.NaN, 4, 5])

fig = plt.figure( 3 )
ax = fig.add_subplot( 111 )
bp = ax.boxplot( [a], vert = True)
plt.show()
```

Removing the np.NaN gives me the desired result

Reply:
OK, so I think you are almost there.  Looks like you will have to do the "pop" trick.  *But*, before you do, I think you need to do `X=X.T` if its a 2-D masked array because the semantics of 2-D arrays are different than a list of lists. 

I would still like to see the test have a 2-D masked array as the input....  i.e. I think: `data = np.array([[1, 2, 3, -1,], [3, 2, -1, 1]])` `a = np.ma.masked_less(data, 0)`

Reply:
Added a test case for the 2-D masked array.